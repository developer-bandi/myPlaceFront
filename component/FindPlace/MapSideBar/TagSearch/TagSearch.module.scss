.mainBlock {
  overflow-y: scroll;
  height: calc(100vh - 111px);
  padding: 20px;
  @media screen and (max-width: 767px) {
    height: calc(100% - 50px);
  }
}

.categoryListBlock {
  display: flex;
  margin: 15px 0 15px 0;
}

@mixin category {
  padding: 5px;
  border-radius: 10px;
  color: black;
  border: 1px solid #e5e5e5;
  font-size: 17px;
  margin-right: 10px;
  padding: 5px 15px 5px 15px;
  &:hover {
    background: #87cefa;
    border: 1px solid #87cefa;
  }
}

.category {
  @include category();
}

.selectedCategory {
  @include category();
  background: #87cefa;
  border: 1px solid #87cefa;
}

.inputBlock {
  margin: 0 0 20px 0;
}

.adressInput {
  width: calc(100% - 60px);
  height: 30px;
  border: 1px solid black;
  border-radius: 5px;
  margin-right: 10px;
  @media screen and (max-width: 767px) {
    width: calc(100vw - 10px);
  }
}

.adressButton {
  width: 50px;
  height: 30px;
  background-color: #00aaff;
  border-radius: 5px;
  color: white;
}

.title {
  font-size: 16px;
  font-weight: bold;
  margin: 10px 0 15px 0;
}

.hashtagListBlock {
  margin-bottom: 15px;
  display: flex;
  flex-wrap: wrap;
}

@mixin hashtag {
  border: 1px solid #e5e5e5;
  border-radius: 10px;
  padding: 7px;
  margin-right: 10px;
  margin-bottom: 10px;
  font-size: 16px;
  cursor: pointer;
  &:hover {
    opacity: 0.6;
  }
}

.hashtag {
  @include hashtag();
}

.selectedHashtag {
  @include hashtag();
  background: #87cefa;
  border: 1px solid #87cefa;
  color: white;
}

.searchbutton {
  display: block;
  width: 100%;
  height: 40px;
  margin: 0 auto 20px auto;
  background-color: #00aaff;
  border-radius: 10px;
  color: white;
  cursor: pointer;
  &:hover {
    opacity: 0.8;
  }
}
