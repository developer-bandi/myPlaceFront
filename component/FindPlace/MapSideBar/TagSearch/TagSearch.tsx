import React, { forwardRef, RefObject, SetStateAction } from "react";
import Category from "./Category/Category";
import Hashtag from "./Hashtag/Hashtag";
import Position from "./Position/Position";
import styles from "./TagSearch.module.scss";

interface TagSearchProps {
  hashtag:
    | {
        [index: string]: {
          [index: string]: [string, number, number][];
        };
      }
    | undefined;
  selectedHashtag: string[];
  selectedCategory: string;
  dispatchCategory: (category: string) => void;
  dispatchHashtag: (hashtag: string) => void;
  dispatchAddress: () => void;
  inputRef: RefObject<HTMLInputElement>;
  dispatchSearchStore: () => void;
}

const TagSearch = ({
  hashtag,
  selectedHashtag,
  selectedCategory,
  dispatchCategory,
  dispatchHashtag,
  dispatchAddress,
  inputRef,
  dispatchSearchStore,
}: TagSearchProps) => {
  return (
    <main className={styles.mainBlock}>
      <Category
        selectedCategory={selectedCategory}
        dispatchCategory={dispatchCategory}
      />
      <Position dispatchAddress={dispatchAddress} inputRef={inputRef} />
      <Hashtag
        hashtag={hashtag}
        selectedCategory={selectedCategory}
        selectedHashtag={selectedHashtag}
        dispatchHashtag={dispatchHashtag}
      />
      <button
        className={styles.searchbutton}
        onClick={() => {
          dispatchSearchStore();
        }}
      >
        검색하기
      </button>
    </main>
  );
};

const categorydata = ["카페", "식당", "주점"];

export default TagSearch;
