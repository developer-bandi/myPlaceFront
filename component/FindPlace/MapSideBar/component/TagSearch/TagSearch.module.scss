.mainBlock {
  height: 860px;
  overflow-y: scroll;
  @media screen and (max-width: 767px) {
    height: 450px;
  }
}

.categoryListBlock {
  display: flex;
  padding-left: 10px;
  margin-top: 15px;
  margin-bottom: 15px;
}

@mixin category {
  padding: 5px;
  border-radius: 10px;
  color: black;
  border: 1px solid #e5e5e5;
  font-size: 17px;
  margin-right: 10px;
  padding: 5px 15px 5px 15px;
  &:hover {
    background: #87cefa;
    border: 1px solid #87cefa;
  }
}

.category {
  @include category();
}

.selectedCategory {
  @include category();
  background: #87cefa;
  border: 1px solid #87cefa;
}

.inputBlock {
  padding-left: 10px;
  margin-bottom: 20px;
}

.adressInput {
  width: 290px;
  height: 30px;
  border: 1px solid black;
  border-radius: 5px;
  margin-right: 10px;
}

.adressButton {
  width: 50px;
  height: 30px;
  background-color: #00aaff;
  border-radius: 5px;
  color: white;
}

.title {
  padding-left: 10px;
  font-size: 16px;
  font-weight: bold;
  margin-bottom: 15px;
  margin-top: 10px;
}

.hashtagListBlock {
  padding-left: 10px;
  margin-bottom: 15px;
  display: flex;
  flex-wrap: wrap;
}

@mixin hashtag {
  border: 1px solid #e5e5e5;
  border-radius: 10px;
  padding: 7px;
  margin-right: 10px;
  margin-bottom: 10px;
  font-size: 16px;
  cursor: pointer;
}

.hashtag {
  @include hashtag();
}

.selectedHashtag {
  @include hashtag();
  background: #87cefa;
  border: 1px solid #87cefa;
  color: white;
}

.searchbutton {
  display: block;
  width: 360px;
  height: 40px;
  margin: auto;
  background-color: #00aaff;
  border-radius: 10px;
  color: white;
  cursor: pointer;
  margin-bottom: 20px;
}
