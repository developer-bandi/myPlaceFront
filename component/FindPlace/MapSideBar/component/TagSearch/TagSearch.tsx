import { forwardRef, RefObject, SetStateAction } from "react";
import styles from "./TagSearch.module.scss";

interface TagSearchProps {
  hashtag: {
    [index: string]: {
      [index: string]: [string, number, number][];
    };
  };
  selectedHashtag: string[];
  selectedCategory: string;
  dispatchCategory: (category: string) => void;
  dispatchHashtag: (hashtag: string) => void;
  dispatchAddress: () => void;
  inputRef: RefObject<HTMLInputElement>;
  dispatchSearchStore: () => void;
}

const TagSearch = ({
  hashtag,
  selectedHashtag,
  selectedCategory,
  dispatchCategory,
  dispatchHashtag,
  dispatchAddress,
  inputRef,
  dispatchSearchStore,
}: TagSearchProps) => {
  return (
    <main className={styles.mainBlock}>
      <div>
        <ul className={styles.categoryListBlock}>
          {categorydata.map((data: string, index: number) => {
            return (
              <li
                className={
                  selectedCategory === data
                    ? styles.selectedCategory
                    : styles.category
                }
                key={data}
                onClick={() => {
                  dispatchCategory(data);
                }}
              >
                {data}
              </li>
            );
          })}
        </ul>
      </div>
      <section className={styles.centerPostisionBlock}>
        <h1 className={styles.title}>기준 장소</h1>
        <div className={styles.inputBlock}>
          <input className={styles.adressInput} ref={inputRef} />
          <button
            className={styles.adressButton}
            onClick={() => {
              dispatchAddress();
            }}
          >
            설정
          </button>
          <div className={styles.alert}></div>
        </div>
      </section>
      <section className={styles.tagscontainer} key={"tagscontainer"}>
        {Object.keys(hashtag[selectedCategory]).map((data, index: number) => {
          return (
            <>
              <div className={styles.title}>{data}</div>
              <ul className={styles.hashtagListBlock} key={index}>
                {hashtag[selectedCategory][data].map((data) => {
                  return (
                    <li
                      className={
                        selectedHashtag.indexOf(data[0]) !== -1
                          ? styles.selectedHashtag
                          : styles.hashtag
                      }
                      key={data[0]}
                      onClick={() => {
                        dispatchHashtag(data[0]);
                      }}
                    >
                      #{data[0]}
                    </li>
                  );
                })}
              </ul>
            </>
          );
        })}
      </section>
      <button
        className={styles.searchbutton}
        onClick={() => {
          dispatchSearchStore();
        }}
      >
        검색하기
      </button>
    </main>
  );
};

const categorydata = ["카페", "식당", "주점"];

export default TagSearch;
