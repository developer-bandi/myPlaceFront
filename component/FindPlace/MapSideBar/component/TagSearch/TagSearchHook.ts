import {useEffect, useRef} from "react";
import {useDispatch, useSelector} from "react-redux";
import {RootReducer} from "../../../../../store";
import {
  setCategory,
  addHashTag,
  deleteHashTag,
  setAdress,
} from "../../../../../store/reducers/hashtagSearchCondition/Reducer";
import {searchStore} from "../../../../../store/reducers/searchResult/Reducer";
import {setSearchType} from "../../../../../store/reducers/searhType/Reducer";

const useTagSearch = () => {
  const dispatch = useDispatch();
  const inputRef = useRef<HTMLInputElement>(null);
  const hashtag = useSelector((state: RootReducer) => state.hashtagAll.content);
  const selectedHashtag = useSelector(
    (state: RootReducer) => state.hashtagSearchCondition.hashtag
  );
  const selectedCategory = useSelector(
    (state: RootReducer) => state.hashtagSearchCondition.category
  );
  const settedAdress = useSelector(
    (state: RootReducer) => state.hashtagSearchCondition.adress
  );

  useEffect(() => {
    if (inputRef.current !== null)
      inputRef.current.value = settedAdress.content;
  }, [settedAdress]);

  const dispatchHashtag = (hashtag: string) => {
    if (selectedHashtag.indexOf(hashtag) == -1) {
      dispatch(addHashTag(hashtag));
    } else {
      dispatch(deleteHashTag(hashtag));
    }
  };

  const dispatchCategory = (category: string) => {
    dispatch(setCategory(category));
  };

  const dispatchAddress = () => {
    if (inputRef.current !== null) {
      dispatch(setAdress({adress: inputRef.current.value}));
    }
  };

  const dispatchSearchStore = () => {
    dispatch(
      searchStore({
        latitude: settedAdress.latitude,
        longitude: settedAdress.longitude,
        selectedHashtag,
      })
    );
    dispatch(setSearchType({type: "hashtag", hashtag: "result"}));
  };

  return {
    hashtag,
    selectedHashtag,
    selectedCategory,
    dispatchCategory,
    dispatchAddress,
    dispatchHashtag,
    inputRef,
    dispatchSearchStore,
  };
};

export default useTagSearch;
