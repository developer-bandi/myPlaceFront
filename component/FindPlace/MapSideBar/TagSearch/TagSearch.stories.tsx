import { ComponentStory, ComponentMeta } from "@storybook/react";
import TagSearch from "./TagSearch";

export default {
  title: "FindPlace/MapSideBar/TagSearch",
  component: TagSearch,
  parameters: {
    layout: "fullscreen",
  },
} as ComponentMeta<typeof TagSearch>;

const Template: ComponentStory<typeof TagSearch> = (arg) => {
  return (
    <div style={{ width: 340, position: "relative" }}>
      <TagSearch {...arg} />
    </div>
  );
};

export const generalType = Template.bind({});
generalType.args = {
  hashtag: { 카페: { 음식: [["아이스아메리카노", 1, 1]] } },
  selectedHashtag: ["아이스아메리카노"],
  selectedCategory: "카페",
};
