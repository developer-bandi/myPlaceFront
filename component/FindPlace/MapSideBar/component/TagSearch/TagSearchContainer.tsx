import TagSearch from "./TagSearch";
import useTagSearch from "./TagSearchHook";

const TagSearchContainer = () => {
  const {
    hashtag,
    selectedHashtag,
    selectedCategory,
    dispatchCategory,
    dispatchAddress,
    dispatchHashtag,
    inputRef,
    dispatchSearchStore,
  } = useTagSearch();
  return (
    <TagSearch
      hashtag={hashtag}
      selectedHashtag={selectedHashtag as string[]}
      selectedCategory={selectedCategory as string}
      dispatchCategory={dispatchCategory}
      dispatchAddress={dispatchAddress}
      dispatchHashtag={dispatchHashtag}
      inputRef={inputRef}
      dispatchSearchStore={dispatchSearchStore}
    ></TagSearch>
  );
};

export default TagSearchContainer;
