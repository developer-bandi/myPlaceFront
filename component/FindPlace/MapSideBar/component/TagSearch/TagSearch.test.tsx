import useTagSearch from "./TagSearchHook";
import {
  act,
  fireEvent,
  render,
  renderHook,
  screen,
} from "@testing-library/react";
import {Provider} from "react-redux";
import configureMockStore from "redux-mock-store";
import newReact from "react";
import * as React from "react";
import TagSearch from "./TagSearch";

const mockStore = configureMockStore()({
  hashtagSearchCondition: {
    adress: {
      content: "testAdress",
      latitude: "testLatitude",
      longitude: "testLongitude",
    },
    keyword: "testKeyword",
    hashtag: ["testHashtag"],
  },
  hashtagAll: {},
});
const wrapper = ({children}: any) => (
  <Provider store={mockStore}>{children}</Provider>
);
describe("TagSearch Hook 테스트", () => {
  it("useEffect 테스트", () => {
    jest
      .spyOn(newReact, "useRef")
      .mockReturnValueOnce({current: {value: "test"}});
    const {result} = renderHook(() => useTagSearch(), {
      wrapper,
    });
    expect(result.current.inputRef.current).toEqual({value: "testAdress"});
  });
  describe("dispatchHashtag 함수 테스트", () => {
    it("추가하는 경우", () => {
      const {result} = renderHook(() => useTagSearch(), {
        wrapper,
      });
      act(() => {
        result.current.dispatchHashtag("testHashtag2");
      });
      expect(mockStore.getActions()[0]).toEqual({
        payload: "testHashtag2",
        type: "hashtagSearchCondition/addHashTag",
      });
    });
    it("삭제하는 경우", () => {
      const {result} = renderHook(() => useTagSearch(), {
        wrapper,
      });
      act(() => {
        result.current.dispatchHashtag("testHashtag");
      });
      expect(mockStore.getActions()[1]).toEqual({
        payload: "testHashtag",
        type: "hashtagSearchCondition/deleteHashTag",
      });
    });
  });
  it("dispatchCategory 함수 테스트", () => {
    const {result} = renderHook(() => useTagSearch(), {
      wrapper,
    });
    act(() => {
      result.current.dispatchCategory("testCategory");
    });
    expect(mockStore.getActions()[2]).toEqual({
      payload: "testCategory",
      type: "hashtagSearchCondition/setCategory",
    });
  });
  it("dispatchAddress 함수 테스트", () => {
    jest
      .spyOn(newReact, "useRef")
      .mockReturnValueOnce({current: {value: "test"}});
    const {result} = renderHook(() => useTagSearch(), {
      wrapper,
    });
    act(() => {
      result.current.dispatchAddress();
    });
    expect(mockStore.getActions()[3]).toEqual({
      payload: {
        adress: "testAdress",
      },
      type: "hashtagSearchCondition/setAdress",
    });
  });
  it("dispatchSearchStore 함수 테스트", () => {
    const {result} = renderHook(() => useTagSearch(), {
      wrapper,
    });
    act(() => {
      result.current.dispatchSearchStore();
    });
    expect(mockStore.getActions()[4]).toEqual({
      payload: {
        latitude: "testLatitude",
        longitude: "testLongitude",
        selectedHashtag: ["testHashtag"],
      },
      type: "searchResult/searchStore",
    });
    expect(mockStore.getActions()[5]).toEqual({
      payload: {
        hashtag: "result",
        type: "hashtag",
      },
      type: "searchType/setSearchType",
    });
  });
});

describe("TagSearch Presentational 테스트", () => {
  it("정상 렌더링", () => {
    const dispatchCategoryMock = jest.fn();
    const dispatchHashtagMock = jest.fn();
    const dispatchAddressMock = jest.fn();
    const dispatchSearchStoreMock = jest.fn();
    const utils = render(
      <TagSearch
        hashtag={{
          카페: {
            목적: [["test", 1, 1]],
          },
        }}
        selectedHashtag={[]}
        selectedCategory={"카페"}
        dispatchCategory={dispatchCategoryMock}
        dispatchHashtag={dispatchHashtagMock}
        dispatchAddress={dispatchAddressMock}
        inputRef={{current: null}}
        dispatchSearchStore={dispatchSearchStoreMock}
      ></TagSearch>
    );
    expect(utils.container).toMatchSnapshot();
    fireEvent.click(screen.getByTestId("dispatchCategory0"));
    fireEvent.click(screen.getByTestId("dispatchHashtag0"));
    fireEvent.click(screen.getByTestId("dispatchAddress"));
    fireEvent.click(screen.getByTestId("dispatchSearchStore"));
    expect(dispatchCategoryMock).toBeCalledTimes(1);
    expect(dispatchHashtagMock).toBeCalledTimes(1);
    expect(dispatchAddressMock).toBeCalledTimes(1);
    expect(dispatchSearchStoreMock).toBeCalledTimes(1);
  });
});
