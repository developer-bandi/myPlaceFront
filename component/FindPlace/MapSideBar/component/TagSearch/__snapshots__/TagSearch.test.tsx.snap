// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagSearch Presentational 테스트 정상 렌더링 1`] = `
<div>
  <main
    class="mainBlock"
  >
    <div>
      <ul
        class="categoryListBlock"
      >
        <li
          class="selectedCategory"
          data-testid="dispatchCategory0"
        >
          카페
        </li>
        <li
          class="category"
          data-testid="dispatchCategory1"
        >
          식당
        </li>
        <li
          class="category"
          data-testid="dispatchCategory2"
        >
          주점
        </li>
      </ul>
    </div>
    <section
      class="centerPostisionBlock"
    >
      <h1
        class="title"
      >
        기준 장소
      </h1>
      <div
        class="inputBlock"
      >
        <input
          class="adressInput"
        />
        <button
          class="adressButton"
          data-testid="dispatchAddress"
        >
          설정
        </button>
        <div
          class="alert"
        />
      </div>
    </section>
    <section
      class="tagscontainer"
    >
      <div
        class="title"
      >
        목적
      </div>
      <ul
        class="hashtagListBlock"
      >
        <li
          class="hashtag"
          data-testid="dispatchHashtag0"
        >
          #
          test
        </li>
      </ul>
    </section>
    <button
      class="searchbutton"
      data-testid="dispatchSearchStore"
    >
      검색하기
    </button>
  </main>
</div>
`;
